### PR TITLE
event: scaled timers using dynamic queues

### PR DESCRIPTION
**Pull request by akonradi**
Commit Message: Add ScaledRangeTimerManager using dynamic queues
Additional Description:
The ScaledRangeTimerManager will be used to implement overload actions
that scale timeouts, as suggested for #11427. The timer objects returned remain
attached to the manager so that their actual trigger points can be adjusted via
changes to the manager's scale factor.

Risk Level: low (unused code)
Testing: ran unit tests
Docs Changes: none
Release Notes: none

#11427
/cc @antoniovicente

### CODE REVIEW

File: test/mocks/event/wrapped_dispatcher.cc
Comment (antoniovicente) [inline test/mocks/event/wrapped_dispatcher.cc:6]: nit: Consider moving the implementation of trivial forwarding functions to the header file.

File: test/common/event/scaled_range_timer_manager_test.cc
Comment (antoniovicente) [inline test/common/event/scaled_range_timer_manager_test.cc:136]: What does pending and active mean in this context?  Seems different from the states encoded in the RangeTimerImpl absl::variant

File: test/common/event/scaled_range_timer_manager_test.cc
Comment (antoniovicente) [inline test/common/event/scaled_range_timer_manager_test.cc:144]: Worth adding some assertions about the queue state after the advance and run?

File: test/common/event/scaled_range_timer_manager_test.cc
Comment (antoniovicente) [inline test/common/event/scaled_range_timer_manager_test.cc:187]: Move this expect call just below this comment.  There's currently a big disconnect between this expectation and comments that refer to it.

// After the original windows for both timers have long expired, only the enabled one should fire.

File: test/common/event/scaled_range_timer_manager_test.cc
Comment (antoniovicente) [inline test/common/event/scaled_range_timer_manager_test.cc:417]: Please add a test that covers use of setScaleFactor.  There seems to be no tests that exercise that function.
Also, prefer use of setScaleFactor instead of passing in a scale value to the constructor.

I recommend exercising some some in-between scale factors like 0.5 with timers that are expected to fire with different orders when scale = 1.0 and scale = 0.5
Cover both the case of timers added before the scale change and timers added after the scale change.  Bonus points for mixing the two.

File: test/common/event/scaled_range_timer_manager_test.cc
Comment (antoniovicente) [inline test/common/event/scaled_range_timer_manager_test.cc:412]: See the comment I added to ScaledRangeTimerManager::onQueueTimerFired

**Review by antoniovicente — COMMENTED** [review]
Thanks Alex, this seems like huge progress towards implementation of overload manager aware timeouts.  Sorry for the laundry list of comments, it sometimes unavoidable on PRs of this size and complexity.

File: source/common/event/scaled_range_timer_manager.cc
Comment (akonradi) [reply source/common/event/scaled_range_timer_manager.cc:189]: Do we want to fire multiple timers as a group? It seems like we can get the same watchdog petting between callback executions by rescheduling the queue's timer with duration 0. I guess that might be extra expensive, though.

Comment (antoniovicente) [reply source/common/event/scaled_range_timer_manager.cc:189]: Timers with 0 duration run on the next iteration of the event loop after https://github.com/envoyproxy/envoy/pull/11823

In order to schedule a second call in the same loop iteration you would need to use to use SchedulableCallback::scheduleCallbackCurrentIteration to schedule in cases where remaining max-min duration is 0.

Comment (antoniovicente) [reply source/common/event/scaled_range_timer_manager.cc:189]: Triggering expired timers as a group seems like a good way to implement this.  We can address the watchdog petting issues in a followup.

File: source/common/event/scaled_range_timer_manager.cc
Comment (akonradi) [reply source/common/event/scaled_range_timer_manager.cc:152]: Because the returned value is a reference to the inserted element, but this needs an iterator.

File: test/common/event/scaled_range_timer_manager_test.cc
Comment (KBaichoo) [reply test/common/event/scaled_range_timer_manager_test.cc:26]: Yes, my idea with the onFatalError extensions would be to invoke them here: https://github.com/envoyproxy/envoy/blob/12a103cb4bd667208cd5fba9b3bf29defe2c1dfe/source/common/event/dispatcher_impl.h#L88

So it currently wouldn't need a public accessor. 

An accessor would get tricky if you try to add an accessor that wasn't running on the same thread, and I don't think it would be a good idea.

File: test/common/event/scaled_range_timer_manager_test.cc
Comment (akonradi) [reply test/common/event/scaled_range_timer_manager_test.cc:418]: MultipleTimersWithScaling does some of this. I've added some more tests.

File: test/common/event/scaled_range_timer_manager_test.cc
Comment (akonradi) [reply test/common/event/scaled_range_timer_manager_test.cc:144]: It's not part of the public interface, so I'm hesitant to test it directly.

File: test/common/event/scaled_range_timer_manager_test.cc
Comment (akonradi) [reply test/common/event/scaled_range_timer_manager_test.cc:136]: Pending was from an old implementation, should be fixed now.

File: source/common/event/scaled_range_timer_manager.cc
Comment (akonradi) [reply source/common/event/scaled_range_timer_manager.cc]: Changed this to trigger multiple timers when a queue timer expires.

File: source/common/event/scaled_range_timer_manager.cc
Comment (akonradi) [reply source/common/event/scaled_range_timer_manager.cc]: Done.

File: source/common/event/scaled_range_timer_manager.h
Comment (akonradi) [reply source/common/event/scaled_range_timer_manager.h]: Left over from a previous implementation. Now private.

File: source/common/event/scaled_range_timer_manager.h
Comment (akonradi) [reply source/common/event/scaled_range_timer_manager.h]: Done.

Comment (akonradi) [reply source/common/event/scaled_range_timer_manager.h]: Fixed.

File: source/common/event/scaled_range_timer_manager.h
Comment (akonradi) [reply source/common/event/scaled_range_timer_manager.h]: Done.

File: test/mocks/event/wrapped_dispatcher.cc
Comment (akonradi) [reply test/mocks/event/wrapped_dispatcher.cc]: They're all trivial, so removed the .cc file.

File: test/common/event/scaled_range_timer_manager_test.cc
Comment (antoniovicente) [inline test/common/event/scaled_range_timer_manager_test.cc:166]: It seems that you want to cover behavior in cases where the timer scaling max which is responsible for the queue timer is disabled, and check that the next timer in the queue takes over.  There are no assertions on changes to the queue timer.

Possible suggestion: Advance time to T=30 and verify that nothing triggers, move to T=35 and verify that a timer fires.

File: test/common/event/scaled_range_timer_manager_test.cc
Comment (antoniovicente) [inline test/common/event/scaled_range_timer_manager_test.cc:406]: Please add comment explaining expected behavior; cases where max < min are treated as if max == min.

**Review by antoniovicente — COMMENTED** [review]
Almost there.  Thanks again for this very exciting change.

File: test/common/event/scaled_range_timer_manager_test.cc
Comment (akonradi) [reply test/common/event/scaled_range_timer_manager_test.cc]: Thanks for the pointer. Added a test.

File: source/common/event/scaled_range_timer_manager.cc
Comment (antoniovicente) [inline source/common/event/scaled_range_timer_manager.cc:31]: Where does the transition from scaling-max to inactive come from?

File: test/common/event/scaled_range_timer_manager_test.cc
Comment (antoniovicente) [inline test/common/event/scaled_range_timer_manager_test.cc:491]: .After is new to me.  Cool stuff!

File: test/common/event/scaled_range_timer_manager_test.cc
Comment (antoniovicente) [inline test/common/event/scaled_range_timer_manager_test.cc:535]: nit: You lost some of the + after the T to start rename.

**Review by antoniovicente — DISMISSED** [review]

File: test/common/event/scaled_range_timer_manager_test.cc
Comment (akonradi) [reply test/common/event/scaled_range_timer_manager_test.cc]: Thanks, fixed.

File: source/common/event/scaled_range_timer_manager.cc
Comment (akonradi) [reply source/common/event/scaled_range_timer_manager.cc:31]: If the timer is already enabled, the first step is to disable it. Then we can enable it again for the new duration.

File: test/common/event/scaled_range_timer_manager_test.cc
Comment (akonradi) [reply test/common/event/scaled_range_timer_manager_test.cc]: 🤦 thanks, fixed.

File: source/common/event/scaled_range_timer_manager.cc
Comment (antoniovicente) [reply source/common/event/scaled_range_timer_manager.cc:31]: I missed that the timer described in the comment is initially in the scaling-max state.  The example may have been easier to follow if the initial state had been waiting-for-min

**Review by antoniovicente — DISMISSED** [review]
Change looks really good.  Thanks for sticking with it until the end, looking forward to softer overload actions that protect the proxy while reducing query loss.

Off to @mattklein123 or @envoyproxy/senior-maintainers for review and merge.

**Comment by akonradi** [comment]
@mattklein123 @htuch this implements the first scheme from the doc, which it sounds like might be sufficient for our needs. Tagging you both since you left feedback on the doc, PTaL.

**Comment by mattklein123** [comment]
Sure happy to take a look.

**Review by htuch — COMMENTED** [review]
I'll defer to others on implementation review, but I'm happy with this after a high-level read over the PR, nice work!

File: source/common/event/scaled_range_timer_manager.h
Comment (mattklein123) [inline source/common/event/scaled_range_timer_manager.h:74]: nit: const, same elsewhere where you can.

File: source/common/event/scaled_range_timer_manager.cc
Comment (mattklein123) [inline source/common/event/scaled_range_timer_manager.cc:173]: Let me try to explain this code back to you to see if I understand it:

1) Every range timer has an individual min timer that fires independently.
2) Once the min timer fires, if there is a gap between min and max, we take the duration of the min/max delta, and put it in a queue unique to that delta. The assumption here is that there will be a small number of min/max configured so the number of queues is not that big.
3) Each queue has a timer that fires for the min/max duration scaled, and flushes all timers that have expired, and then resets for the next duration that needs to fire.

If so that makes sense, but per my other comments can you add a bunch more comments in these two files?

**Review by mattklein123 — CHANGES_REQUESTED** [review]
I just reviewed the code. Overall LGTM with a few small mostly comment comments. Thanks!

/wait

File: source/common/event/scaled_range_timer_manager.cc
Comment (akonradi) [reply source/common/event/scaled_range_timer_manager.cc:187]: Yep, that's correct. Added plenty of comments.

**Review by mattklein123 — COMMENTED** [review]
Awesome, thanks for the comments. LGTM modulo format fix.

/wait

**Review by mattklein123 — APPROVED** [review]

**Comment by akonradi** [comment]
Looks like the failing test is `//test/extensions/transport_sockets/tls:handshaker_test` which is flaky under Windows.

**Pull request merged**
